commands/src/main/java/org/wildfly/extras/creaper/commands/security/realms/AddLdapConnection.java

```diff
@@ -0,0 +1,212 @@
+        this.url = builder.url;
+        this.replaceExisting = builder.replaceExisting;
+        this.properties = builder.properties;
+    }
+
+    @Override
+    public void apply(OnlineCommandContext ctx) throws CliException, CommandFailedException, IOException,
+            TimeoutException, InterruptedException {
+        Operations ops = new Operations(ctx.client);
+        Address ldapConnection = Address.coreService("management").and("ldap-connection", connectionName);
+        if (replaceExisting) {
+            try {
+                ops.removeIfExists(ldapConnection);
+                new Administration(ctx.client).reloadIfRequired();
+            } catch (OperationException e) {
+                throw new IOException("Failed to LDAP outbound connection " + connectionName, e);
+            }
+        }
+        List<String> resolvedHandlesReferralsFor = (handlesReferralsFor != null && !handlesReferralsFor.isEmpty())
+                ? handlesReferralsFor
+                : null;
+        ops.add(ldapConnection, Values.empty()
+                .andOptional("name", connectionName)
+                .andOptional("handles-referrals-for", initialContextFactory)
+                .andOptional("initial-context-factory", initialContextFactory)
+                .andOptional("referrals", referrals)
+                .andOptional("search-credential", searchCredential)
+                .andOptional("search-dn", searchDn)
+                .andOptional("security-realm", securityRealm)
+                .andOptional("url", url)
+                .andListOptional(String.class, "handles-referrals-for", resolvedHandlesReferralsFor));
+
+        if (properties != null && !properties.isEmpty()) {
+            for (Map.Entry<String, String> entry : properties.entrySet()) {
+                ops.add(ldapConnection.and("property", entry.getKey()), Values.empty()
+                        .and("value", entry.getValue())
+                );
+            }
+        }
+    }
+
+    @Override
+    public void apply(OfflineCommandContext ctx) throws CommandFailedException, IOException {
+        Map<String, Object> groovyAttributes = new HashMap<String, Object>();
+        groovyAttributes.put("atrConnectionName", connectionName);
+        groovyAttributes.put("atrHandlesReferralsFor", handlesReferralsFor);
+        groovyAttributes.put("atrInitialContextFactory", initialContextFactory);
+        groovyAttributes.put("atrReferrals", referrals);
+        groovyAttributes.put("atrSearchCredential", searchCredential);
+        groovyAttributes.put("atrSearchDn", searchDn);
+        groovyAttributes.put("atrSecurityRealm", securityRealm);
+        groovyAttributes.put("atrUrl", url);
+        groovyAttributes.put("atrReplaceExisting", replaceExisting);
+        groovyAttributes.put("atrProperties", properties);
+
+        ctx.client.apply(GroovyXmlTransform.of(AddLdapConnection.class)
+                .parameters(groovyAttributes)
```

Comment:
Not sure why you don't use `parameter` here, like this:

```java
ctx.client.apply(GroovyXmlTransform.of(AddLdapConnection.class)
        .parameter("atrConnectionName", connectionName)
        .parameter("atrHandlesReferralsFor", handlesReferralsFor)
```

Not that building a `Map` upfront is bad, but for this straightforward case, I think it doesn't make much sense.